File: backend/people/relation_vocab.py
"""Allowed relation modifiers and cores for Person (server validation)."""

from __future__ import annotations

RELATION_PREFIX_TOKENS: frozenset[str] = frozenset(
    {
        "great",
        "step",
        "half",
        "adopted",
        "adoptive",
        "distant",
        "god",
        "younger",
        "older",
        "foster",
        "twin",
        "triplet",
    }
)

RELATION_SUFFIX_TOKENS: frozenset[str] = frozenset({"in_law", "best"})

RELATION_CORE_CHOICES: tuple[tuple[str, str], ...] = (
    ("self", "Self"),
    ("mother", "Mother"),
    ("father", "Father"),
    ("aunt", "Aunt"),
    ("uncle", "Uncle"),
    ("niece", "Niece"),
    ("nephew", "Nephew"),
    ("cousin", "Cousin"),
    ("spouse", "Spouse"),
    ("partner", "Partner"),
    ("significant_other", "Significant other"),
    ("brother", "Brother"),
    ("sister", "Sister"),
    ("grandpa", "Grandpa"),
    ("grandma", "Grandma"),
    ("friend", "Friend"),
    ("pet", "Pet"),
    ("child", "Child"),
    ("son", "Son"),
    ("daughter", "Daughter"),
)

RELATION_CORE_VALUES: frozenset[str] = frozenset(c for c, _ in RELATION_CORE_CHOICES)


def default_alias_for_core(core: str) -> str:
    return dict(RELATION_CORE_CHOICES).get(core, core.replace("_", " "))


def validate_prefix_tokens(tokens: list | None) -> list[str]:
    if not tokens:
        return []
    if not isinstance(tokens, list):
        raise ValueError("relation_prefix_tokens must be a list.")
    out: list[str] = []
    for t in tokens:
        s = str(t).strip().lower()
        if not s:
            continue
        if s not in RELATION_PREFIX_TOKENS:
            raise ValueError(f"Unknown relation prefix token: {t!r}.")
        out.append(s)
    return out


def validate_suffix_tokens(tokens: list | None, *, relation_core: str) -> list[str]:
    if not tokens:
        return []
    if not isinstance(tokens, list):
        raise ValueError("relation_suffix_tokens must be a list.")
    out: list[str] = []
    for t in tokens:
        s = str(t).strip().lower()
        if not s:
            continue
        if s not in RELATION_SUFFIX_TOKENS:
            raise ValueError(f"Unknown relation suffix token: {t!r}.")
        if s == "best" and relation_core != "friend":
            raise ValueError("Suffix 'best' is only allowed when relation_core is 'friend'.")
        out.append(s)
    return out

File: frontend/src/people/relationVocab.ts
/** Mirrors backend `people/relation_vocab.py` for form validation and labels. */

export const RELATION_PREFIX_TOKENS: { value: string; label: string }[] = [
  { value: "great", label: "Great" },
  { value: "step", label: "Step" },
  { value: "half", label: "Half" },
  { value: "adopted", label: "Adopted" },
  { value: "adoptive", label: "Adoptive" },
  { value: "distant", label: "Distant" },
  { value: "god", label: "God" },
  { value: "younger", label: "Younger" },
  { value: "older", label: "Older" },
  { value: "foster", label: "Foster" },
  { value: "twin", label: "Twin" },
  { value: "triplet", label: "Triplet" },
];

export const RELATION_CORE_SELF = { value: "self", label: "Self" } as const;

export const RELATION_CORE_OPTIONS: { value: string; label: string }[] = [
  { value: "mother", label: "Mother" },
  { value: "father", label: "Father" },
  { value: "brother", label: "Brother" },
  { value: "sister", label: "Sister" },
  { value: "child", label: "Child" },
  { value: "son", label: "Son" },
  { value: "daughter", label: "Daughter" },
  { value: "aunt", label: "Aunt" },
  { value: "uncle", label: "Uncle" },
  { value: "niece", label: "Niece" },
  { value: "nephew", label: "Nephew" },
  { value: "cousin", label: "Cousin" },
  { value: "spouse", label: "Spouse" },
  { value: "partner", label: "Partner" },
  { value: "significant_other", label: "Significant other" },
  { value: "grandpa", label: "Grandpa" },
  { value: "grandma", label: "Grandma" },
  { value: "friend", label: "Friend" },
  { value: "pet", label: "Pet" },
];

export function relationCoreSelectOptions(currentCore: string): { value: string; label: string }[] {
  if (currentCore === "self") {
    return [RELATION_CORE_SELF, ...RELATION_CORE_OPTIONS];
  }
  return RELATION_CORE_OPTIONS;
}

export function labelForRelationCore(core: string): string {
  if (core === "self") return RELATION_CORE_SELF.label;
  return RELATION_CORE_OPTIONS.find((o) => o.value === core)?.label ?? core;
}
